src/library/scala/collection/immutable/LinkedMap.scala

```diff
@@ -0,0 +1,209 @@
+package scala
+package collection
+package immutable
+
+import java.io.{ObjectInputStream, ObjectOutputStream}
+
+import scala.collection.mutable.{Builder, ImmutableBuilder}
+
+/**
+  * A generic trait for ordered immutable maps. Concrete classes have to provide
+  * functionality for the abstract methods in `LinkedMap`:
+  *
+  * @tparam A
+  * @tparam B
+  */
+
+trait LinkedMap[K, +V]
+  extends Map[K, V]
+    with Iterable[(K, V)]
+    with MapOps[K, V, LinkedMap, LinkedMap[K, V]]
+    with Equals
+
+object LinkedMap extends MapFactory[LinkedMap] {
+  def empty[K, V]: LinkedMap[K, V] = EmptyLinkedMap.asInstanceOf[LinkedMap[K, V]]
+
+  def from[K, V](it: collection.IterableOnce[(K, V)]): LinkedMap[K, V] =
+    it match {
+      case vm: LinkedMap[K, V] => vm
+      case _ => (newBuilder[K, V] ++= it).result()
+    }
+
+  def newBuilder[K, V]: Builder[(K, V), LinkedMap[K, V]] =
+    new ImmutableBuilder[(K, V), LinkedMap[K, V]](empty) {
+      def addOne(elem: (K, V)): this.type = { elems = elems + elem; this }
+    }
+
+  @SerialVersionUID(3L)
+  private object EmptyLinkedMap extends LinkedMap[Any, Nothing] {
+    override def size: Int = 0
+    override def knownSize: Int = 0
+    override def apply(key: Any) = throw new NoSuchElementException("key not found: " + key)
+    override def contains(key: Any) = false
+    def get(key: Any): Option[Nothing] = None
+    override def getOrElse [V1](key: Any, default: => V1): V1 = default
+    def iterator: Iterator[(Any, Nothing)] = Iterator.empty
+    def updated [V1] (key: Any, value: V1): LinkedMap[Any, V1] = new LinkedMap1(key, value)
+    def remove(key: Any): LinkedMap[Any, Nothing] = this
+  }
+
+  @SerialVersionUID(3L)
+  final class LinkedMap1[K, +V](key1: K, value1: V) extends LinkedMap[K,V] with Serializable {
```

Comment:
This should have been `private`, along with all the other ...N's, IMO.

Reply:
Ah this must have been overlooked, should I make a PR to fix this?

Reply:
@mdedetrich That would be awesome!